File: services/networkx_service/src/process/metric/profiler.py
import os
import gc
import time
import psutil
import multiprocessing
from threading import Thread
from pubsub.user_publisher import UserPublisher
from pubsub.algorithms_data_publisher import AlgorithmsDataPublisher


class Profiler:
    def __init__(self, socket_id, file_id, algorithm_name, algorithm_type):
        self.socket_id = socket_id
        self.file_id = file_id
        self.algorithm_name = algorithm_name
        self.algorithm_type = algorithm_type

        self.initialized = False
        self.initialized_memory_set = False
        self.stopped = False
        self.emitted_data_points = 1

        self.total_memory = 0
        self.total_cpu = 0
        self.last_recorded_time = None
        self.initialization_memory = 0

        self.start_time = None

    def get_memory(self):
        process = psutil.Process(os.getpid())
        return process.memory_info().rss / 1024 / 1024

    def get_cpu(self):
        process = psutil.Process(os.getpid())
        return process.cpu_percent(interval=0.1) / multiprocessing.cpu_count()

    def get_time(self):
        return time.time() - self.start_time

    def __emit(self):
        current_memory = self.get_memory() - self.initialization_memory
        current_cpu = self.get_cpu()
        current_time = self.get_time()

        if self.initialized and current_memory >= 0:
            self.total_memory += current_memory
            self.total_cpu += current_cpu
            self.emitted_data_points += 1

        if self.initialized and (not self.initialized_memory_set or current_memory < 0):
            self.initialization_memory = self.get_memory()
            self.initialized_memory_set = True

            if current_memory < 0:
                current_memory = self.initialization_memory
            else:
                current_memory = 0

        self.last_recorded_time = current_time

        user_data = {
            "event_name": "metric_emit",
            "payload": {
                "algorithm_name": self.algorithm_name,
                "metrics": {
                    "memory": current_memory,
                    "cpu": current_cpu,
                },
                "time": current_time
            }
        }

        UserPublisher(self.socket_id, self.algorithm_name).send(user_data)

    def __post_emit(self):
        user_data = {
            "event_name": "metric_end",
            "payload": {
                "algorithm_name": self.algorithm_name,
                "time": self.last_recorded_time
            }
        }

        UserPublisher(self.socket_id, self.algorithm_name).send(user_data)

        data_aggregator_features = {
            "avg_memory": self.total_memory / self.emitted_data_points,
            "avg_cpu": self.total_cpu / self.emitted_data_points,
            "total_time": self.last_recorded_time
        }

        AlgorithmsDataPublisher(self.file_id, self.algorithm_name, self.algorithm_type).send(data_aggregator_features)

    def __monitor(self):
        self.start_time = time.time()

        while not self.stopped:
            try:
                self.__emit()
            except Exception as e:
                self.stopped = True
                break

        self.__post_emit()

    def mark_initialization(self):
        gc.collect()
        gc.disable()
        self.initialized = True

    def start(self):
        Thread(target=self.__monitor).start()

    def stop(self):
        self.stopped = True

File: services/jgrapht_service/src/process/src/main/java/process/metric/Profiler.java
package process.metric;

import java.lang.management.ManagementFactory;
import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.TimeUnit;

import com.sun.management.OperatingSystemMXBean;
import org.json.JSONObject;
import process.pubsub.AlgorithmsDataPublisher;
import process.pubsub.Publisher;
import process.pubsub.UserPublisher;

public class Profiler {
    private final String socketId;
    private final String fileId;
    private final String algorithmName;
    private final String algorithmType;

    private final Runtime runtime;
    private final OperatingSystemMXBean os;

    volatile boolean initialized = false;
    volatile boolean initializedMemorySet = false;
    volatile boolean stopped = false;
    volatile long emittedDataPoints = 1;
    volatile double totalMemory = 0;
    volatile double totalCpu = 0;
    volatile double lastRecordedTime = 0;
    volatile double initializationMemory = 0;
    Instant startTime;

    private final Publisher userPublisher;
    private final Publisher algorithmPublisher;

    public Profiler(String socketId, String fileId, String algorithmName, String algorithmType) {
        this.socketId = socketId;
        this.fileId = fileId;
        this.algorithmName = algorithmName;
        this.algorithmType = algorithmType;

        this.runtime = Runtime.getRuntime();
        this.os = ManagementFactory.getPlatformMXBean(com.sun.management.OperatingSystemMXBean.class);

        this.userPublisher = new UserPublisher(socketId, algorithmName);
        this.algorithmPublisher = new AlgorithmsDataPublisher(fileId, algorithmName, algorithmType);
    }

    private double getMemory() {
        return (double) (runtime.totalMemory() - runtime.freeMemory()) / (1024L * 1024L);
    }

    private double getCpu() {
        return os.getCpuLoad() * 100;
    }

    private double getTime() {
        Instant currentTime = Instant.now();

        return (double) Duration.between(startTime, currentTime).toMillis() / 1000;
    }

    private synchronized void emit() {
        double currentMemory = getMemory() - initializationMemory;
        double currentCpu = getCpu();
        double currentTime = getTime();

        if (initialized && currentMemory >= 0) {
            totalMemory += currentMemory;
            totalCpu += currentCpu;
            emittedDataPoints += 1;
        }

        if (initialized && (!initializedMemorySet || currentMemory < 0)) {
            initializationMemory = getMemory();
            initializedMemorySet = true;

            if (currentMemory < 0) {
                currentMemory = initializationMemory;
            } else {
                currentMemory = 0;
            }
        }

        lastRecordedTime = currentTime;

        JSONObject userData = new JSONObject()
                .put("event_name", "metric_emit")
                .put("payload", new JSONObject()
                        .put("algorithm_name", algorithmName)
                        .put("metrics", new JSONObject()
                                .put("memory", currentMemory)
                                .put("cpu", currentCpu)
                        )
                        .put("time", currentTime)
                );

        userPublisher.send(userData);
    }

    private void postEmit() {
        JSONObject userData = new JSONObject()
                .put("event_name", "metric_end")
                .put("payload", new JSONObject()
                        .put("algorithm_name", algorithmName)
                        .put("time", lastRecordedTime)
                );

        userPublisher.send(userData);

        JSONObject aggregatorData = new JSONObject()
                .put("avg_memory", totalMemory / emittedDataPoints)
                .put("avg_cpu", totalCpu / emittedDataPoints)
                .put("total_time", lastRecordedTime);

        algorithmPublisher.send(aggregatorData);
    }

    private void monitor() {
        startTime = Instant.now();

        while (!stopped) {
            try {
                TimeUnit.MILLISECONDS.sleep(100);
                emit();
            } catch (InterruptedException e) {
                stopped = true;
                e.printStackTrace();
            }
        }

        postEmit();
    }

    public void markInitialization() {
        System.gc();
        initialized = true;
    }

    public Thread start() {
        Thread thread = new Thread(this::monitor);
        thread.start();

        return thread;
    }

    public void stop() {
        stopped = true;
    }
}
